### ISSUE TITLE
How linux readerex differs from windows readerex in font usage

### ISSUE THREAD

**Issue by HinTak** [caj2pdf/caj-convert#2]
- linux version: does not embed in either case, but use different non-embedded named fonts to draw, depending on whether simsun is present or not.

- windows version: does not draw text if font is missing. Texts silently missing if fonts are not present.

caj2pdf#63 works correctly in wine against the windows dlls if simsun is coped from windows into wine's font directory.

**Comment by HinTak** [comment]
Went through all the dll/libs - the windows ordinal 216 is actually stable. But:

- dll from 2.3.3979.0 onwards can embed all fonts; 2.3.3978.0 is like the linux one, not embedding a few fonts.
- 32-bit windows is more reliable; 64-bit seems to have indeterminisms (repeat runs give slightly different outcome).
- versioning meaningless - 4 versions of 2.3.3982 gives 4 different outcome; the two 32-bit builds have different dependencies.
- the outcome differs by dependencies more than by versions (79-82 vs 82-83 differs ). And small version change in 78-79 gives major differences.
- in terms of font embedding, the linux version almost behaves like older windows ones.

About ordinal 216 - it works okay for issue-63 (other than font embedding) upto 2.0.3949.0, then 2.0.3920.0 breaks for this sample, and seems to not have jbig (just jbig2 + jpeg). 2.0.3684.0 and 2.0.3799.0 can read the jpeg cover only.
Older versions cannot read NH files.

**Comment by ghost** [comment]
Explained in code. If no font can be found, defaultFont is used to be an alternative to `报宋`, following the definition below.

```cpp
namespace GlobalParams
{
#if defined(OS_LINUX) || defined(OS_MACOS)
const char *defaultFontName = "HGHT_CNKI";
#elif defined(OS_WINDOWS)
const char *defaultFontName = "SimSun";
#endif
} // namespace GlobalParams
```
```cpp
if (!strcmp(FontName, "报宋"))
{
	FontName = GlobalParams::defaultFontName;
	bFlag = 0;
}
```

**Comment by HinTak** [comment]
Is that your code or cnki's? Seems incorrect, as per my early comment (linux version DOES try simsun):

linux version: does not embed in either case, but use different non-embedded named fonts to draw, depending on whether simsun is present or not.

**Comment by Adam-Gao** [comment]
No, it is just a code snippet. The snippet itself is okay, and you can use `gdb` to confirm this.

 The following string references may give you some hints.
```python
CAJDocEditor::ProcessText(CAJPage *,WITS_21_S72_DRAWATTR &,CCmdObj *,void *,void *)
	"报宋"

GetBuziFont(char const*,int &)
	"buzi_%02X.ttf"

CAJDocEditor::SetZhaozi(HI_WITS_21_S72 &,std::ostream *,int,char const*)
	"【E】无法找到补字字体，%s, %s, 第%d页，字体，%02X，字符编码，%02X"
	"【E】无法找到补字字体，%s, 使用默认字体代替，字体编码, %02X"
	"【E】不能加载补字信息，%s, 第%d页，字体，%02X，字符编码，%02X"
	"【E】不能加载补字信息, %s, 第%d页, 字体编码, %02X, 字符编码, %02X"
```
